newIDE/app/src/ResourcesList/ResourcePreview/Resource3DPreviewContext.js

```diff
@@ -0,0 +1,268 @@
+          );
+          if (pendingErrorPromise) {
+            pendingErrorPromise.resolve(this.fallbackImagePath);
+            this.pendingPromises.delete(errorResourceUrl);
+          }
+          break;
+        default:
+          console.warn('Unknown message type from worker:', type);
+          break;
+      }
+    };
+
+    this.worker.onerror = error => {
+      console.error('Worker error:', error);
+      // Resolve any pending promises with the fallback image
+      this.pendingPromises.forEach(promise => {
+        promise.resolve(this.fallbackImagePath);
+      });
+      this.pendingPromises.clear();
+    };
+  }
+
+  initWorker() {
+    const message: WorkerInitMessage = { type: MESSAGE_TYPES.INIT };
+    this.worker.postMessage(message);
+  }
+
+  renderModel(resourceUrl: string): Promise<string> {
+    return new Promise((resolve, reject) => {
+      if (!this.isInitialized) {
+        resolve(this.fallbackImagePath);
+        return;
+      }
+
+      this.pendingPromises.set(resourceUrl, { resolve, reject });
+      const message: WorkerRenderModelMessage = {
+        type: MESSAGE_TYPES.RENDER_MODEL,
+        resourceUrl,
+      };
+      this.worker.postMessage(message);
+    });
+  }
+
+  terminate() {
+    if (this.worker) {
+      this.worker.terminate();
+    }
+  }
+}
+
+type Props = {|
+  children: React.Node,
+|};
+
+export const Resource3DPreviewProvider = ({ children }: Props) => {
+  const [currentResource, setCurrentResource] = React.useState<?string>(null);
+  const queueRef = React.useRef<
+    Array<{ url: string, resolve: (dataUrl: ?string) => void }>
+  >([]);
+  const previewCache = React.useRef<{ [url: string]: string }>({});
+  const workerManagerRef = React.useRef<?Resource3DPreviewWorkerManager>(null);
+
+  // Initialize the worker manager on mount
+  React.useEffect(() => {
+    workerManagerRef.current = new Resource3DPreviewWorkerManager();
+
+    // Cleanup on unmount
+    return () => {
+      queueRef.current = [];
+      previewCache.current = {};
+
+      // Terminate the worker
+      if (workerManagerRef.current) {
+        workerManagerRef.current.terminate();
+        workerManagerRef.current = null;
+      }
+    };
+  }, []);
+
+  const enqueueResource = React.useCallback((url: string): Promise<?string> => {
+    return new Promise(resolve => {
+      // If it's already in the cache, resolve immediately.
+      if (previewCache.current[url]) {
+        resolve(previewCache.current[url]);
+        return;
+      }
+
+      // Add the item to the queue.
+      queueRef.current.push({ url, resolve });
+      // If the queue didn't have items before,
+      // then process it immediately.
+      // Otherwise, let the queue process handle it.
+      if (queueRef.current.length === 1) {
+        setCurrentResource(url);
+      }
+    });
+  }, []);
+
+  // Process the next resource in the queue
+  const processNextResource = React.useCallback(async (url: string) => {
```

Comment:
Nitpicking: this function does process a "next" resource. It has no knowledge that what you pass is the "next" or not - it just process the url.
So it could be just "renderResource" (because render is more specific) or event "renderModel" (because it's what it does, it does not support any other resource)